src/deepmol/datasets/datasets.py

```diff
@@ -225,32 +239,39 @@ def __init__(self,
         if not isinstance(mols, np.ndarray):
             mols = np.array(mols)
         if not isinstance(X, np.ndarray) and X is not None:
-            X = np.ndarray(X)
+            X = np.array(X)
         if not isinstance(y, np.ndarray) and y is not None:
-            y = np.ndarray(y)
+            y = np.array(y)
         if not isinstance(ids, np.ndarray) and ids is not None:
-            ids = np.ndarray(ids)
+            ids = np.array([str(i) for i in ids])
         if not isinstance(features2keep, np.ndarray) and features2keep is not None:
-            features2keep = np.ndarray(features2keep)
+            features2keep = np.array(features2keep)
 
-        self.mols = mols
+        self._mols = mols
 
         if features2keep is not None:
-            self.features2keep = features2keep
+            self._features2keep = features2keep
+        elif X is not None:
+            if len(X.shape) == 1:
+                self._features2keep = np.arange(X.shape[0])
+            else:
+                self._features2keep = np.arange(X.shape[1])
         else:
-            self.features2keep = None
+            self._features2keep = None
 
-        self.X = X
-        self.y = y
-        self.ids = ids
-        self.n_tasks = n_tasks
+        self._X = X
+        self._y = y
+        if ids is None:
+            ids = np.array([str(uuid.uuid4().hex) for i in range(len(mols))])
```

Comment:
By the way, wouldn't you save this line of code if you simply put **self.ids = ids**? As I saw in the **setter** you already do this verification... so it seems you are repeating the code here

Reply:
Hum, I didn't think much about performance... I believe they have slightly worse performance than regular IDs. Still, if we don't have extremely strict memory requirements it is ok to use them because they offer other advantages like reducing the chances of having repeated IDs which will highly impact other methods in our class.

Reply:
Regarding the second part, I understand what you are saying, and probably things could be improved but for now, in the constructor, I set the internal version of ids (`self._ids`) so the setter will not be called. This way the setter will only be used outside the class call (`NumpyDataset.ids = ids`) not when instantiating the class itself (`NumpyDataset(..., ids=ids`).

Should I rethink this?

Reply:
No, I think it is a reasonable explanation. For me it s good